### PR TITLE
NEW: Add stainaugment feature for stain augmentation

### PR DESCRIPTION
**Pull request by mostafajahanifar**
Introducing `stainaugment` tool for generating stain-augmented images from an input image. This class can be used stand-alone or in albumentations augmentation pipelines.

The process works as bellow:
1. H and E parts of the input images are first extracted, using a predefined `stain_matrix` or a `stain_matrix` that has been calculated based on 'Vahadane' or 'Macenko' methods.
2. Input image's stain concentrations are calculated based on the provided/extracted `stain_matrix`.
3. Stain concentrations are adjusted randomly based on two input `sigma1` and `sigma2` parameters:
    * Scaled by `alpha = np.random.uniform(1 - self.sigma1, 1 + self.sigma1)`
    * Biased by `beta = np.random.uniform(-self.sigma2, self.sigma2)`
4. New image is formed based on the adjusted concentrations.

Providing the pre-extracted `stain_matrix` on class initialization have two benefits:
- `stainaugment` can be used for small patches on-the-fly during the model training (because the bulk of stain matrix extraction or dictionary learning is eliminated)
- By setting the parameters, this function allow the user to augment the stain around a target stain matrix i.e., normalize the stain to a target image and then augment it.

- [x] Class definition and structure
- [x] Check the functionality
- [x] Add support for albumentations 
- [x] Complete the docstring and writing examples
- [x] Complete the test and coverage
- [ ] ~~Add an example to "Stain Normalization" notebook~~

### CODE REVIEW

File: tests/test_stainaugment.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+def test_stainaugment(source_image, norm_vahadane):
+    """Test functionality of the StainAugmentation class."""
+    source_img = imread(pathlib.Path(source_image))
+    target_img = stainnorm_target()
+    vahadane_img = imread(pathlib.Path(norm_vahadane))
+
+    # Test invalid method in the input
+    with pytest.raises(ValueError, match=r".*Unsupported stain extractor method.*"):
+        _ = StainAugmentation(method="mosi")
```
Comment (simongraham) [inline tests/test_stainaugment.py:23]: I like how you put your nickname, but shall we put something more clear like ‘invalid’

File: tests/test_stainaugment.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+    # 1. Testing without stain matrix.
+    # Test with Vahadane stain extractor
+    augmentor = StainAugmentation(
```
Comment (simongraham) [inline tests/test_stainaugment.py:27]: You said testing with Vahadane but this is Macenko

File: tiatoolbox/tools/stainaugment.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+# Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+#
+# This file contains code inspired by StainTools
+# [https://github.com/Peter554/StainTools] written by Peter Byfield.
+#
+# The Original Code is Copyright (C) 2021, TIACentre, University of Warwick
+# All rights reserved.
+# ***** END GPL LICENSE BLOCK *****
+
+"""Staing augmentation"""
```
Comment (simongraham) [inline tiatoolbox/tools/stainaugment.py:24]: Should be ‘Stain augmentation’

File: tiatoolbox/tools/stainaugment.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+            (`beta` belonging to [-sigma2, sigma2] range). Default is 0.25.
+        augment_background (bool): Specifies whether to apply stain augmentation on the
+            background or not. Default is False, which indicates that only tissue region
+            will be stain augmented.
+        always_apply (False): For use with 'albumentations' pipeline. Please refer to
+            albumentations documentations for more information.
+        p (0.5): For use with 'albumentations' pipeline which specifies the probability
+            of using the augmentation in a 'albumentations' pipeline. . Please refer to
+            albumentations documentations for more information.
+    Attributes:
```
Comment (simongraham) [inline tiatoolbox/tools/stainaugment.py:66]: Should put space before Attributes, Examples etc

**Review by simongraham — APPROVED** [review]
Nice work. Few small things to change but generally it’s nice. Happy to merge when those things are addressed

File: tests/test_stainaugment.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+import numpy as np
+import pytest
+
+from tiatoolbox.data import stainnorm_target
+from tiatoolbox.tools.stainaugment import StainAugmentation
+from tiatoolbox.tools.stainnorm import get_normaliser
+from tiatoolbox.utils.misc import imread
+
+
+def test_stainaugment(source_image, norm_vahadane):
```
Comment (Srijay-lab) [inline tests/test_stainaugment.py:15]: If the function is testing with Vahadane only, why can't we name it as test_vahadane_stainaugment() ?

File: tests/test_stainaugment.py (hunk 1 above)
Comment (mostafajahanifar) [reply tests/test_stainaugment.py:23]: thanks for seeing this :-D I forget to change it.

File: tests/test_stainaugment.py (hunk 5 above)
Comment (mostafajahanifar) [reply tests/test_stainaugment.py:15]: Both Vahadane and Macenko methods are tested in the function, however, only files `norm_vahadane` and `source_image` are used in some part of the test.

File: tiatoolbox/tools/stainaugment.py (hunk 3 above)
Comment (mostafajahanifar) [reply tiatoolbox/tools/stainaugment.py]: Thanks Simon.
All your comments have been considered in commit ff0e65c.

**Comment by mostafajahanifar** [comment]
Many thanks tp @simongraham and @Srijay-lab for reviewing this PR. I have incorporated your comments in ff0e65c. Please let us know if you are happy with the changes (and test the code on your system if possible) so @shaneahmed can merge this PR.

**Review by Srijay-lab — APPROVED** [review]

File: tiatoolbox/tools/stainaugment.py [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+        """Return an augmented instance based on source stain concentrations.
+
+        Stain concentrations of the source image are altered (scaled and shifted)
+        based on the random alpha and beta paramters, and then an augmented image is
+        reconstructed from the altered concentrations.
+        All parameters needed for this part are calculated when calling fit() function.
+
+        Returns:
+            img_augmented (:class:`numpy.ndarray`): stain augmented image.
+        """
```
Comment (shaneahmed) [inline tiatoolbox/tools/stainaugment.py:174]: need an extra line.

File: tiatoolbox/tools/stainaugment.py [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+        to extract stain matrix and stain concentrations of the input image to be used
+        in the `augment` function.
+
+        Args:
+            img (:class:`numpy.ndarray`): RGB image in the form of uint8 numpy array.
+            threshold (float): The threshold value used to find tissue mask from the
+                luminosity component of the image. The found `tissue_mask` will be used
+                to filter out background area in stain augmentation process upon user
+                setting `augment_background=False`.
+        """
```
Comment (shaneahmed) [inline tiatoolbox/tools/stainaugment.py:151]: empty line

File: tiatoolbox/tools/stainaugment.py [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+        >>> stain_matrix = np.array([[0.91633014, -0.20408072, -0.34451435],
+        ...                [0.17669817, 0.92528011, 0.33561059]])
+        >>> # Instantiate the stain augmentor and fit it on an image
+        >>> stain_augmentor = StainAugmentation(stain_matrix=stain_matrix)
+        >>> stain_augmentor.fit(img)
+        >>> # Now using the fitted `stain_augmentor` in a loop to generate
+        >>> # several augmented instances from the same image.
+        >>> for i in range(10):
+        ...     img_aug = stain_augmentor.augment()
+    """
```
Comment (shaneahmed) [inline tiatoolbox/tools/stainaugment.py:104]: new line

**Review by shaneahmed — CHANGES_REQUESTED** [review]
Thanks @mostafajahanifar 
Please address my comments and add stain augmentation feature to `usage.rst`

File: tiatoolbox/tools/stainaugment.py (hunk 8 above)
Comment (mostafajahanifar) [reply tiatoolbox/tools/stainaugment.py:105]: all docstring issues are addressed in 26dcf3f.
Thanks Shan.

**Review by shaneahmed — APPROVED** [review]
Thanks @mostafajahanifar

**Pull request merged**